src/Hazelcast.Net/Networking/SslLayer.cs

```diff
@@ -65,9 +66,42 @@ public async ValueTask<Stream> GetStreamAsync(Stream stream)
 
             if (!_options.Enabled) return stream;
 
-            var sslStream = new SslStream(stream, false, ValidateCertificate, null);
+            var clientCertificates = GetClientCertificates();
 
-            var clientCertificates = GetClientCertificatesOrDefault();
+            LocalCertificateSelectionCallback? selectCertificate = null;
+            // note: OperatingSystem.IsWindows() is n/a in netstandard
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && clientCertificates.Count > 1)
+            {
+                // on e.g. linux if the collection of certificates contains more than 1 certificate, then the SslStream
+                // is going to fail selecting the correct one + it's going to fail sending the chain to the server, so
+                // we have to do some housekeeping and workaround here
+
+                // assume there's only going to be 1 cert with a private key, and pick it
+                // should we work with more than 1 PK certs, we'd need a more complex logic
+                selectCertificate = (sender, host, certificates, certificate, issuers) 
+                    => certificates.Cast<X509Certificate2>().FirstOrDefault(x => x.HasPrivateKey);
+
+                // for the whole cert chain to be sent as part of the handshake we need to copy the
+                // certs that don't have a PK to the current user CA certificate store
+                var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser);
+                try
+                {
+                    store.Open(OpenFlags.ReadWrite);
+                    foreach (var cert in clientCertificates)
+                    {
+                        _logger.IfDebug()?.LogDebug($"CertFix: {cert.Subject}   (issuer: {cert.Issuer})   {cert.Thumbprint}   [{(cert.HasPrivateKey ? "PK" : "ADD")}]");
+                        if (!cert.HasPrivateKey) store.Add(cert);
```

Comment:
Here, given certificates are added to the store. Is it a permanent action on current user's store? If yes, I think having a warning message could be appropriate.

Reply:
Makes sense, will do. I also notice that the test coverage is decreasing with this PR, going to look into it.